### PR TITLE
Restructuring

### PR DESCRIPTION
**Pull request by akosyakov**

### CODE REVIEW

**Comment by svenefftinge** [comment]
Nice! For others, it would be good if you could provide an overview what you did and what the purpose is.

**Comment by svenefftinge** [comment]
Related to #59

**Comment by akosyakov** [comment]
It breaks down theia-core on core and filesystem extensions and shows how they are used together for the browser example. It does not relies on ldm anymore but use lerna for linking submodules with hoisting dependencies and 2 tsconfigs, one to improve navigation in vscode, another to compile efficiently. Because of linking debugging works out of the box, without adjusting tsconfigs or generated source maps.

I am going to:
- enable tests again
- restructure remaining extensions and examples
- add tslint rules for imports, one should always use `@theia/${extension}/lib/*` imports to import something from other extensions
- adapt app generators and a new extension generator
- update development docs

**Comment by hexa00** [comment]
Does that sound like a good time to also change to absolute paths ? 
Since you'll have to rewrite the relatives ones anyway

File: core/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "theia-core",
-  "version": "0.0.1-alpha.2",
+  "name": "@theia/core",
+  "version": "0.0.1-alpha.1",
   "description": "Theia is a cloud & desktop IDE framework implemented in TypeScript.",
   "repository": {
     "type": "git",
```
Comment (hexa00) [inline core/package.json]: This is back to .1 ?

Comment (svenefftinge) [inline core/package.json]: I agree, we should keep the version even though we change the package name.

**Comment by svenefftinge** [comment]
You mean use absolute paths even within an npm package?

**Comment by hexa00** [comment]
I did not realise at first that segmenation would do that by default with the different packages.
Sounds like it would be nide also inside the same package but I'm not sure what is involed there or what is best pratice?

File: core/package.json (hunk 1 above)
Comment (akosyakov) [inline core/package.json]: ok

**Comment by akosyakov** [comment]
I've tried to use the absolute paths inside the same package, e.g. in core `@theia/core/lib/*`, and rolled back because they are getting long. But what is nice about them that you can copy imports around and move files without breaking imports often.

**Comment by akosyakov** [comment]
It is generally done. @svenefftinge @hexa00 Please review. Could you update the docs as you try it out?

There are following follow up tasks left:
- [ ] enable examples' tests
- [ ] split the filesystem ext to get rid of cycle with the preferences ext
- [ ] fixed monaco integration after switching to 0.9.0
- [ ] add import rules for tslint
- [ ] shrinkwrap is not going to work because of hoisting => extraneous dependencies, try with npm5 lock files instead

**Comment by svenefftinge** [comment]
Thanks Anton, I will look into it tomorrow.

**Comment by hexa00** [comment]
Same here most likely tomorrow

**Comment by svenefftinge** [comment]
updating to the latest phospor breaks the layout.
See https://github.com/theia-ide/theia/issues/30

**Comment by svenefftinge** [comment]
We should remove the 'localDependencies' in the example's theia-package.json.

**Comment by svenefftinge** [comment]
Otherwise , it worked nicely for me.
Incremental development was easy and fast. :)

**Comment by akosyakov** [comment]
@svenefftinge I've addressed issues which you mention and update docs.

**Comment by hexa00** [comment]
I wish you had waited a bit more before merging this ?
I don't understand why this was merged without the examples for example and their ci for example ?
Sounds like we're missing some stuff compared to before....
Or I'm missing something

**Comment by akosyakov** [comment]
Examples are there, only UI tests are switched off for now. I am looking into switching them on.

**Comment by hexa00** [comment]
Yeah I mean their ci but ok I just think we could have waiting to get this in the PR before
I think we should avoid breaking existing stuff in master

**Comment by hexa00** [comment]
Note also the TLDR doc is not broken 
should be cd examples/browser rather than cd ../../examples/browser

**Comment by akosyakov** [comment]
ok, I will update it

**Comment by akosyakov** [comment]
@hexa00 I've tried to get it done over the weekend that there are no merge issues. Unfortunately, there was one which I was not able to rebase already because of tree changes and had to merge. There are some small tasks left, I will clean it up next days.

**Comment by akosyakov** [comment]
I am a bit concern about UI tests, they don't seem to do much but takes so long. Mabe better to have a separate job for them after examples are successfully build.

**Comment by akosyakov** [comment]
Speed up is because of dependencies are downloaded once for all packages and tasks are executed in parallel there is possible.

**Comment by hexa00** [comment]
Nice speed up for sure, the colors look nice too hehe :)

**Comment by kittaakos** [comment]
> the colors look nice too

👍

**Comment by hexa00** [comment]
I think for the UI tests if we start a job after it will be the same as having it in the same job.

If ever our core test become long we could have 2 job in parallel one executing only the core tests and the other the UI tests only. And wait for the 2 to complete.

But I think right now it's the install/compile that is taking time... I'll have to recheck with the restructuring

**Pull request merged**